zmq4_routerdealer_test.go

```diff
@@ -239,3 +240,35 @@ func TestRouterDealer(t *testing.T) {
 		})
 	}
 }
+
+func TestRouterClose(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	router := zmq4.NewRouter(ctx)
+	err := router.Listen("tcp://*:*")
+	if err != nil {
+		t.Errorf("want error %v, got %q", nil, err)
+	}
+	_, port, _ := net.SplitHostPort(router.Addr().String())
+	dealer := zmq4.NewDealer(ctx)
+	err = dealer.Dial("tcp://*:" + port)
+	if err != nil {
+		t.Errorf("want error %v, got %q", nil, err)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		_, err := router.Recv()
```

Comment:
to be sure we try to receive from a closed router, shouldn't we have _e.g._ another `WaitGroup` that's tickled after `router.Close()` and `Wait()` on it before the `router.Recv()`?

I assume that's what we want to test here.

Reply:
Currently if the socket is closed, `Recv()` is interrupted, returning empty `Frames` and error `nil`. Even if we had that `WaitGroup` the user would still get empty `Frames` and error `nil`. With the fix, `Recv` returns an error so that the user can handle it accordingly (e.g: process frames, exit goroutine...).

Reply:
BTW, it is not only a `router` thing, it is the same in [msgio.go:92](https://github.com/go-zeromq/zmq4/blob/master/msgio.go#L92) and [pub.go:198](https://github.com/go-zeromq/zmq4/blob/master/pub.go#L198).

Reply:
yep. that's something that would need fixing.
(possibly as another PR. or part of this one. as you prefer)

Reply:
I would close the PR as is and do the others separately

Reply:
ok. but could you address the couple of nit-picks I've highlighted?
(then I'll merge)

thanks.